File: Python/Data Structures & Algorithms Module - Python/Week7/MatchBraces.py
"""
MatchBraces.py
Billy Cussen
18/11/2020
"""
import timeit

def matchBraces(word):
    wordLength = len(word)
    if wordLength % 2 != 0:
        return False
    else:
        match = True
        tempLen = wordLength-1
        for i in range(int(wordLength/2)):
            if word[i] != '{' and word[tempLen] != '}':
                match = False
                break
            else:
                tempLen-=1
        return match

def matchBracesRecursion(word):
    if len(word) == 0:
        return True
    
    if len(word) % 2 != 0:
        return False
    elif word[0] == '{' and word[len(word)-1] == '}':
        return matchBracesRecursion(word[1:len(word)-1])
    else:
        return False

startTime = timeit.default_timer()
print("Match Braces: "+str(matchBraces("{{{}}}")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Looped Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer()
print("Match Braces Recursion: "+str(matchBracesRecursion("{{{}}}")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Recursive Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer()
print("Match Braces: "+str(matchBraces("{{{}}")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Looped Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer()
print("Match Braces Recursion: "+str(matchBracesRecursion("{{{}}")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Recursive Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer()
print("Match Braces: "+str(matchBraces("{}")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Looped Approach took "+str(nanoTime)+" nanoseconds")

startTime = timeit.default_timer()
print("Match Braces Recursion: "+str(matchBracesRecursion("{}")))
endTime = timeit.default_timer()
nanoTime = round((endTime-startTime)*1000000,2)
print("Recursive Approach took "+str(nanoTime)+" nanoseconds")

File: Java/Data Structures & Algorithms Module - Java/Week8/Revision/Week7/MatchBraces.java
/*
MatchBraces.java
Billy Cussen
18/11/2020
*/
public class MatchBraces{

    public static void main(String[]args){

        long startTime = System.nanoTime();
        System.out.println("Match Braces ({{{}}}): "+matchBraces("{{{}}}"));
        long endTime = System.nanoTime();
        System.out.println("Looped Approach took "+(endTime-startTime)+" nanoseconds");
        
        startTime = System.nanoTime();
        System.out.println("Match Braces Recursion ({{{}}}): "+matchBracesRecursion("{{{}}}"));
        endTime = System.nanoTime();
        System.out.println("Recursive Approach took "+(endTime-startTime)+" nanoseconds");

        startTime = System.nanoTime();
        System.out.println("Match Braces ({{{}}): "+matchBraces("{{{}}"));
        endTime = System.nanoTime();
        System.out.println("Looped Approach took "+(endTime-startTime)+" nanoseconds");

        startTime = System.nanoTime();
        System.out.println("Match Braces Recursion ({{{}}): "+matchBracesRecursion("{{{}}"));
        endTime = System.nanoTime();
        System.out.println("Recursive Approach took "+(endTime-startTime)+" nanoseconds");

        startTime = System.nanoTime();
        System.out.println("Match Braces ({}): "+matchBraces("{}"));
        endTime = System.nanoTime();
        System.out.println("Looped Approach took "+(endTime-startTime)+" nanoseconds");

        startTime = System.nanoTime();
        System.out.println("Match Braces Recursion ({}): "+matchBracesRecursion("{}"));
        endTime = System.nanoTime();
        System.out.println("Recursive Approach took "+(endTime-startTime)+" nanoseconds");
    }

    public static boolean matchBraces(String braces){
        if(braces.length()%2!=0){
            return false;
        } else{
            boolean checkBraces = true;
            int stringLength = braces.length()-1;
            for(int i = 0; i < braces.length()/2;i++){
                if(braces.charAt(i)=='{' &&braces.charAt(stringLength)=='}'){
                    stringLength--;
                } else{
                    checkBraces = false;
                    break;
                }
            }
            return checkBraces;
        }
    }

    public static boolean matchBracesRecursion(String braces){
        if(braces.length()<=0){
            return true;
        } else if (braces.length()%2!=0){
            return false;
        } 
        
        if(braces.charAt(0)=='{' && braces.charAt(braces.length()-1)=='}'){
            return matchBracesRecursion(braces.substring(1,braces.length()-1));
        } else{
            return matchBracesRecursion(braces.substring(1,braces.length()-1));
        }
    }
}